src/charm.py

```diff
@@ -42,25 +44,72 @@ def __init__(self, *args):
         self.framework.observe(self.on.update_status, self._on_update_status)
         self._provide_mysql()
 
+    def _mysql_layer(self):
+        """Construct the pebble layer"""
+        logger.debug("Building pebble layer")
+        layer = {
+            "summary": "MySQL layer",
+            "description": "Pebble layer configuration for MySQL",
+            "services": {
+                "mysql": {
+                    "override": "replace",
+                    "summary": "mysql service",
+                    "command": "docker-entrypoint.sh mysqld",
+                    "startup": "enabled",
+                    "environment": self.env_config,
+                },
+            },
+        }
+
+        return layer
+
     def _on_peer_relation_joined(self, event):
         if not self.unit.is_leader():
             return
 
-        event.relation.data[event.app][
+        event.relation.data[self.app][
             "MYSQL_ROOT_PASSWORD"
         ] = self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"]
         logger.info("Storing MYSQL_ROOT_PASSWORD in relation data")
+        # logger.warning(event.relation.data[event.app]["MYSQL_ROOT_PASSWORD"])
 
     def _on_peer_relation_changed(self, event):
         if event.relation.data[event.app].get("MYSQL_ROOT_PASSWORD"):
             self._stored.mysql_setup[
                 "MYSQL_ROOT_PASSWORD"
             ] = event.relation.data[event.app]["MYSQL_ROOT_PASSWORD"]
             logger.info("Storing MYSQL_ROOT_PASSWORD in StoredState")
+            # logger.warning(self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"])
+
+    def _on_config_changed(self, event):
+        """Set a new Juju pod specification"""
+        self.unit.status = WaitingStatus("Setting up containers.")
+        container = self.unit.get_container("mysql")
+
+        try:
+            container.add_layer("mysql", self._mysql_layer(), combine=True)
+        except Exception as e:
+            # FIXME: Use a custom Exception
+            logger.debug(e)
+            event.defer()
```

Comment:
Is this surrounded by `try-except` because `config_changed` fires before `pebble_ready`?

Reply:
Sometimes you may [not have the password](https://github.com/canonical/mysql-operator/pull/44/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR213) in the `_stored` because it depends on the [`_on_peer_relation_changed`](https://github.com/canonical/mysql-operator/pull/44/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR76) to be executed